File: backend/qrcode-service/src/main/java/com/quisin/qrcode/controller/QRCodeController.java
package com.quisin.qrcode.controller;

import com.quisin.qrcode.dto.QRCodeResponse;
import com.quisin.qrcode.dto.TableRequest;
import com.quisin.qrcode.dto.TableResponse;
import com.quisin.qrcode.service.QRCodeService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequiredArgsConstructor
@RequestMapping("/tables")
public class QRCodeController {

    private final QRCodeService qrCodeService;

    @PostMapping
    public ResponseEntity<TableResponse> createTable(@Valid @RequestBody TableRequest request) {
        return ResponseEntity.ok(qrCodeService.createTable(request));
    }

    @GetMapping("/restaurant/{restaurantId}")
    public ResponseEntity<List<TableResponse>> getTablesByRestaurantId(@PathVariable String restaurantId) {
        return ResponseEntity.ok(qrCodeService.getTablesByRestaurantId(restaurantId));
    }

    @GetMapping("/restaurant/{restaurantId}/table/{tableNumber}")
    public ResponseEntity<TableResponse> getTableByNumber(
            @PathVariable String restaurantId,
            @PathVariable String tableNumber) {
        return ResponseEntity.ok(qrCodeService.getTableByNumber(restaurantId, tableNumber));
    }

    @GetMapping("/restaurant/{restaurantId}/table/{tableNumber}/qrcode")
    public ResponseEntity<byte[]> generateQRCode(
            @PathVariable String restaurantId,
            @PathVariable String tableNumber,
            @RequestParam(defaultValue = "PNG") String format) {
        
        QRCodeResponse qrCode = qrCodeService.generateQRCode(restaurantId, tableNumber, format);
        
        String contentType = format.equalsIgnoreCase("PDF") ? 
                MediaType.APPLICATION_PDF_VALUE : 
                MediaType.IMAGE_PNG_VALUE;
        
        String filename = String.format("qrcode_%s_%s.%s", 
                restaurantId, 
                tableNumber, 
                format.toLowerCase());

        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType(contentType))
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + filename + "\"")
                .body(qrCode.getQrCodeImage());
    }

    @GetMapping("/restaurant/{restaurantId}/suggest-number")
    public ResponseEntity<String> suggestNextTableNumber(
            @PathVariable String restaurantId,
            @RequestParam String pattern) {
        return ResponseEntity.ok(qrCodeService.suggestNextTableNumber(restaurantId, pattern));
    }

    @DeleteMapping("/restaurant/{restaurantId}/table/{tableNumber}")
    public ResponseEntity<Void> deleteTable(
            @PathVariable String restaurantId,
            @PathVariable String tableNumber) {
        qrCodeService.deleteTable(restaurantId, tableNumber);
        return ResponseEntity.noContent().build();
    }

    @PutMapping("/restaurant/{restaurantId}/table/{tableNumber}")
    public ResponseEntity<TableResponse> updateTable(
            @PathVariable String restaurantId,
            @PathVariable String tableNumber,
            @Valid @RequestBody TableRequest request) {
        return ResponseEntity.ok(qrCodeService.updateTable(restaurantId, tableNumber, request));
    }
} 

File: backend/qrcode-service/src/main/kotlin/com/quisin/qrcode/controller/QRCodeController.kt
package com.quisin.qrcode.controller

import com.quisin.qrcode.service.QRCodeGenerationService
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/qr")
class QRCodeController(
    private val qrCodeGenerationService: QRCodeGenerationService
) {
    @PostMapping("/table")
    fun generateTableQRCode(
        @RequestParam(required = false) tableNumber: String? = null,
        @RequestParam(defaultValue = "300") width: Int = 300,
        @RequestParam(defaultValue = "300") height: Int = 300,
        @RequestParam(defaultValue = "PNG") format: QRCodeGenerationService.OutputFormat = QRCodeGenerationService.OutputFormat.PNG
    ): ResponseEntity<ByteArray> {
        // Use provided table number or generate a new one
        val finalTableNumber = tableNumber ?: qrCodeGenerationService.generateUniqueTableNumber()

        // Generate QR Code
        val qrCodeBytes = qrCodeGenerationService.generateTableQRCode(
            QRCodeGenerationService.QRCodeGenerationOptions(
                tableNumber = finalTableNumber,
                width = width,
                height = height,
                format = format
            )
        )

        // Prepare response headers
        val headers = HttpHeaders().apply {
            contentType = when (format) {
                QRCodeGenerationService.OutputFormat.PNG -> MediaType.IMAGE_PNG
                QRCodeGenerationService.OutputFormat.PDF -> MediaType.APPLICATION_PDF
            }
            contentDisposition = org.springframework.http.ContentDisposition.builder("attachment")
                .filename("table_${finalTableNumber}_qr.${format.name.lowercase()}")
                .build()
        }

        return ResponseEntity(qrCodeBytes, headers, HttpStatus.OK)
    }

    @GetMapping("/generate-table-number")
    fun generateTableNumber(): ResponseEntity<String> {
        return ResponseEntity.ok(qrCodeGenerationService.generateUniqueTableNumber())
    }
}
